cfg/builder/builder_finalize.cc

```diff
@@ -130,92 +131,102 @@ void CFGBuilder::sanityCheck(core::Context ctx, CFG &cfg) {
     }
 }
 
-core::LocalVariable maybeDealias(core::Context ctx, core::LocalVariable what,
-                                 UnorderedMap<core::LocalVariable, core::LocalVariable> &aliases) {
-    if (what.isSyntheticTemporary(ctx)) {
-        auto fnd = aliases.find(what);
+LocalRef maybeDealias(core::Context ctx, CFG &cfg, LocalRef what, UnorderedMap<LocalRef, LocalRef> &aliases) {
+    if (what.isSyntheticTemporary(ctx, cfg)) {
+        auto fnd = aliases.find(what.id());
         if (fnd != aliases.end()) {
             return fnd->second;
-        } else {
-            return what;
         }
-    } else {
-        return what;
     }
+    return what;
 }
 
 /**
  * Remove aliases from CFG. Why does this need a separate pass?
  * because `a.foo(a = "2", if (...) a = true; else a = null; end)`
  */
 void CFGBuilder::dealias(core::Context ctx, CFG &cfg) {
-    vector<UnorderedMap<core::LocalVariable, core::LocalVariable>> outAliases;
-
+    vector<UnorderedMap<LocalRef, LocalRef>> outAliases;
     outAliases.resize(cfg.maxBasicBlockId);
+
     for (auto it = cfg.forwardsTopoSort.rbegin(); it != cfg.forwardsTopoSort.rend(); ++it) {
         auto &bb = *it;
         if (bb == cfg.deadBlock()) {
             continue;
         }
-        UnorderedMap<core::LocalVariable, core::LocalVariable> &current = outAliases[bb->id];
+        auto &current = outAliases[bb->id];
+
         if (!bb->backEdges.empty()) {
+            // Take the intersection of all of the back edges' aliases.
             current = outAliases[bb->backEdges[0]->id];
-        }
-
-        for (BasicBlock *parent : bb->backEdges) {
-            UnorderedMap<core::LocalVariable, core::LocalVariable> other = outAliases[parent->id];
-            for (auto it = current.begin(); it != current.end(); /* nothing */) {
-                auto &el = *it;
-                auto fnd = other.find(el.first);
-                if (fnd != other.end()) {
-                    if (fnd->second != el.second) {
-                        current.erase(it++);
+            for (auto i = 1; i < bb->backEdges.size(); i++) {
```

Comment:
is it the case that back-edges from loops will never be the first back-edge? If so, this would be a great invariant to document here.

Reply:
I don't quite follow. I skip the first backedge because I already include it in the assignment on the previous line.

Reply:
Ah! Sorry, I missed that -- the comment above and the fact that it wasn't part of the diff made it seem like your change was to only process back edges other than the first one.